[웹개발 3주차 미션] 구현 완료했습니다. 리뷰 부탁드립니다.

saint6839 opened the pull request:
개인적으로 바쁜 한 주를 보내서 늦은 리뷰요청 드립니다..!

### 몰랐던 점

- HTML 요청은 GET, POST만 지원을 한다

  다른 HTTP 메서드를 사용하려면 `<input name="_method" type="hidden" value="PUT">` 와 같은 문장을 추가해주면  된다. 그러나 단순히 HTML에 위 코드만 추가했을 경우 여전히 요청이 DELETE가 아닌 POST로 요청됨을 확인하였고, `SpringConfig.class`에 `HiddenHttpMethodFilter ` 객체를 생성해줌으로써 해결할 수 있었다. 

  > `HiddenHttpMethodFilter` :  HiddenHttpMethodFilter는 Hidden 타입의 input 태그의 속성들을 읽어서 HttpServletRequestWrapper.getMethod() 반환 값을 변경해 요청된 HTTP 메소드의 타입을 PUT, DELETE, PATCH로 변경해주는 필터이다.

- ModelAndView

  ModelAndView 타입을 사용해줌으로써 뷰와, 데이터를 함께 저장할 수 있다. Model과의 차이점은 Model은 `addAttribute()`를 이용하여 데이터만 저장하는 반면, ModelAndView는 `addObject()`, `setViewName()`을 이용하여 데이터와, 뷰 페이지를 함께 저장할 수 있다.

- Iterable

  기존에 스프링 데이터 JPA에서 `findAll()`을 호출 하였을 때 반환되었던 `List` 타입과는 다르게, CrudRepository에서는 `Iterable<T>`이 반환되어 리스트 형식인 인자에 사용하기 위해서는 어떻게 해야하나 고민을 했었는데, 결론적으로는 `addAttribute()`의 리스트타입 파라미터에 별도의 형변환 없이 그대로 사용할 수 있었다.

saint6839 committed:
refactor: build.gradle 중복선언제거

saint6839 committed:
refactor: LocalDate, LocalTime -> LocalDateTime으로 변경

saint6839 committed:
refactor: 매핑 url 수정

saint6839 committed:
refactor: method chained call 설정 및 코드 포매팅

saint6839 committed:
refactor: Restful 하게 url 매핑 변경
-PostMapping -> PutMapping
-/users/{id}/update -> /users/{id}

saint6839 committed:
refactor: 메서드명 수정 및 코드 리포매팅

saint6839 committed:
refactor: 메서드명 수정

saint6839 committed:
refactor: Question이 날짜시간을 String이 아닌 LocalDateTime을 가지도록 수정

saint6839 committed:
feat: h2 DB 연동

saint6839 committed:
refactor: userId -> id를 인자로 받도록 수정

saint6839 committed:
refactor: 엔티티로 변경

saint6839 committed:
refactor: 엔티티로 변경 및 유저정보 수정 비밀번호 검증 로직 추가

saint6839 committed:
feat: CrudRepository 추가

saint6839 committed:
refactor: Repository 연동

-UserRepository @AutoWired
-showUserProfileForm : ModelAndView를 반환하도록 수정

saint6839 committed:
refactor: Repository 연동

-QuestionRepository @Autowired
-remove() : 질문 제거 메서드 추가

saint6839 committed:
feat: 유저 정보 수정을 위한 Dto 추가

saint6839 committed:
feat: 유저 정보 수정을 위한 Dto 추가

saint6839 committed:
refactor: path 인자로 id를 받도록 수정

saint6839 committed:
refactor: 사용자 아이디 변경이 불가능하도록 수정

hyukjin-lee commented on src/main/java/codesquad/config/SpringConfig.java:
```diff
@@ -0,0 +1,15 @@
+public class SpringConfig {
+
+    @Bean
+    public HiddenHttpMethodFilter httpMethodFilter() {
```
이 bean 을 꼭 추가해야할까요?

saint6839 replied:
`HiddenHttpMethodFilter`는 hidden 타입의 input 태그 속성을 읽어서 기존에 GET, POST 요청만 사용 가능한 HTML에서 벗어나, 요청된 HTTP 메서드 타입을 PUT, DELETE, PATCH으로 변경해주는 필터라고 공부했습니다.

해당 Bean을 생성하지 않고 프로그램을 실행시켰을 경우 계속해서 PUT 요청이 POST로 보내지는것을 확인해서, 구글링하여 추가해주게 되었습니다..!

hyukjin-lee replied:
HiddenFilter 는 WebMvcConfigurer -> WebMvcAutoConfiguration 에 기본적으로 정의가 되어있는 것으로 기억을 하는데 
가물가물하네요 ~ 한번 확인해보시면 좋을 것 같아요

saint6839 replied:
미션에서는 `WebMvcConfigurer`를 확장하는 `WebMvcConfigurerAdpater`를 사용해서 Config 클래스를 정의하라고 하였지만 deprecated 되었기 떄문에, `WebMvcConfigurer`를 사용해 Config 클래스를 구현하였습니다.

말씀해주신대로 `WebMvcAutoConfiguration`를 확인해보니 `HiddenHttpMethodFilter`가 정의되어 있습니다! 그러나 `WebMvcConfigurer`은 인터페이스이며, `WebMvcAutoConfiguration`은 클래스로 서로간에 확장,구현관계를 가지고 있지는 않은 것 같습니다.

때문에 `WebMvcConfigurer`를 확장해서 Config 클래스를 구성했음에도 해당 인터페이스에는 `HiddenHttpMethodFilter`가 정의되어 있지 않기 떄문에, hidden 타입의 input 태그 속성을 읽지 못해서 별도로 정의해주어야 하는 것 같다고 생각했습니다...!

hyukjin-lee commented on src/main/java/codesquad/controller/QuestionController.java:
```diff
@@ -2,56 +2,52 @@
+
+    private final QuestionRepository questionRepository;
+
+    public QuestionController(QuestionRepository questionRepository) {
```
생성자 주입을 사용한 이유가 있을까요?

saint6839 replied:
생성자 주입은 필드주입이나 세터주입에 비해서 여러가지 장점이 있다고 합니다. 세터 주입은 실제로 객체 변경 가능성때문에 잘 사용하지 않으며, 필드 주입과 생성자 주입은 빈을 주입하는 순서에서 그 차이가 있다고 합니다.

생성자 주입은 객체 생성 시점에 필요한 빈을 찾아서 주입합니다. 때문에 예를 들어 순환 참조일 경우 서로가 참조할 객체가 생성되지 않은 상태에서는 그 객체의 빈을 찾을 수 없으므로 컴파일 에러가 발생합니다.
그러나 필드 주입의 경우에는 빈이 먼저 생성된 후에 필드에 주입하기 때문에, 정상적으로 실행이 되고 그 후에 문제가 발생하기 때문에 오류 파악이 생성자 주입에 비해 늦어집니다.

또한 필드에 final로 선언할 수 있기 때문에 런타임 시점에 객체 불변성이 보장되며, 필드 주입이나 세터 주입으로 빈이 주입되어 있는 경우에는 Mockito를 이용해 모킹을 진행한 후 테스트를 해야하지만,  생성자 주입의 경우에는 단순히 원하는 객체를 생성해서 넣어주기만 하면 되기 때문에 테스트도 간편해집니다.

생성자 주입이 막연하게 좋다고는 알고 있었지만, 이번 기회에 다시 한번 검색해보면서 어떤 부분에서 장점을 띄는지 더 확실하게 알게 된 것 같습니다.

hyukjin-lee replied:
HiddenFilter 는 WebMvcConfigurer -> WebMvcAutoConfiguration 에 기본적으로 정의가 되어있는 것으로 기억을 하는데 
가물가물하네요 ~ 한번 확인해보시면 좋을 것 같아요

hyukjin-lee commented on src/main/java/codesquad/controller/QuestionController.java:
```diff
@@ -2,56 +2,52 @@
-                .orElseThrow(NoSuchQuestionException::new);
+    @GetMapping("/questions/{id}")
+    public String showQuestionDetail(@PathVariable Long id, Model model) {
+        Question question = questionRepository.findById(id)
```
에러는 어떻게 처리되나요?

saint6839 replied:
다음 리뷰때 예외 처리까지 해보도록 하겠습니다!

hyukjin-lee commented on src/main/java/codesquad/exception/NoSuchPasswordException.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.NoSuchElementException;
+
+public class NoSuchPasswordException extends NoSuchElementException {
```
NoSuchElementException 을 사용하지않고 커스텀 exception 을 정의한 이유가 무엇일까요?

saint6839 replied:
커스텀 예외로 의미를 분명하게 하면, 코드 상에서 어떤 예외가 던져지는지 한 눈에 알 수 있지 않을까라는 생각에서 커스텀 예외 클래스를 생성해 보았습니다..!

hyukjin-lee replied:
NoSuchPassword, NoSuchName, NoSuch~~~... etc 로 발전할 가능성이 있어보이는데
장점도 있지만 단점도 있기 때문에 유지보수가 오히려 더 힘들어지는 과한 커스터마이징이 아닌지 고려해보시면 좋을 것 같습니다 ~

hyukjin-lee commented on src/main/java/codesquad/repository/QuestionRepository.java:
```diff
@@ -0,0 +1,7 @@
+import codesquad.domain.Question;
+import org.springframework.data.repository.CrudRepository;
+
+public interface QuestionRepository extends CrudRepository<Question, Long> {
```
Crud vs Jpa Repository 차이점이 뭘까요?

saint6839 replied:
JpaRepository의 경우에는 CrudRepository의 CRUD 기능과 PagingAndSortingRepository의 페이징 및 소팅 기능과 JPA 관련 특화 기능을 사용할 수 있는 인터페이스라고 합니다.

단순 CRUD만 진행하는 프로젝트의 경우에는 CrudRepository를 사용해도 좋으며, 그 외의 페이징, 소팅, jpa 기능을 이용하기 위해서는 JpaRepository 인터페이스를 사용하면 된다고 합니다!

hyukjin-lee commented on src/main/resources/application.properties:
```diff
@@ -3,13 +3,15 @@ handlebars.cache=false
+spring.jpa.show-sql=true
 logging.level.org.hibernate.type.descriptor.sql.BasicBinder=TRACE
+
+spring.jpa.hibernate.ddl-auto=create-drop
```
이 옵션의 종류는 뭐가있고 각각 의미가 무엇일까요?

saint6839 replied:
create : 기존 테이블 삭제 후 재 생성
create-drop : create와 같지만 종료시 테이블을 drop함
update : 변경분만 반영함
validate : 엔티티와 테이블이 잘 매핑되었는지만 확인함
none : 사용하지 않음

개발 초기 단계에서는 create 또는 update를 사용한다
테스트 서버에서는 update 또는 validate를 사용한다.
스테이징과 운영서버는 validate 또는 none을 사용한다.

정도로 찾아보았습니다!

hyukjin-lee reviewed (approved):
고생하셨습니다 ~

saint6839 committed:
refactor: JpaRepository로 변경

saint6839 committed:
feat : 로그인 기능 추가

saint6839 committed:
refactor: 포매팅 수정

saint6839 committed:
refactor: 매핑 경로 수정

saint6839 committed:
refactor: 매핑 경로 수정

saint6839 committed:
refactor: 매핑 경로 수정 및 세션 조건에 따른 문법추가

saint6839 committed:
refactor: css, js 파일 경로 수정

saint6839 committed:
refactor: /users/login registry 제거

saint6839 committed:
feat: UserControllerTest 추가

-회원가입() : 회원가입을 테스트하는 메서드
-회원목록_조회() : 여러 회원이 가입했을떄 리스트의 크기가 일치하는지 테스트하는 메서드
-회원정보_수정() : 가입한 회원 정보가 정상적으로 수정되는지 테스트하는 메서드
-로그인() : 가입한 회원의 로그인 세션정보가 일치하는지 확인하는 메서드

saint6839 committed:
feat: 작가명으로 조회하는 메서드 추가

saint6839 committed:
refactor: 회원가입 테스트 성공하도록 수정

saint6839 committed:
feat: QuestionControllerTest 추가

-질문_등록() : 질문을 등록하고 잘 등록되는지 테스트
-질문목록_조회() : 질문 여러개가 등록 되었을때 리스트의 크기가 일치하는지 테스트
-질문_삭제() : 질문을 등록하고 삭제했을때 잘 삭제되는지 테스트

saint6839 committed:
feat: 로그인 테스트 용이하도록 하기 위해 미리 더미 데이터 생성

saint6839 committed:
refactor: 쿼리문으로 생성한 더미데이터로 로그인 테스트 가능하도록 수정

saint6839 committed:
refactor: 쿼리문으로 생성한 더미데이터 반영해서 테스트 코드 수정

saint6839 committed:
refactor: 쿼리문으로 생성한 더미데이터 반영해서 테스트 코드 수정

saint6839 committed:
refactor: html 수정

saint6839 committed:
refactor: 각 경우에 따른 폼 추가

saint6839 committed:
refactor: 글쓴이 작성란 제거

saint6839 committed:
refactor: 글쓴이 이름 폼에 반영되도록 수정

saint6839 committed:
refactor: 경로 중복 안되도록 registry 경로 제거

saint6839 committed:
feat : 질문 수정 Dto 추가

saint6839 committed:
refactor : 작성한 글쓴이와 로그인한 사용자가 일치할 경우에만 폼 띄워지도록 수정

saint6839 committed:
refactor : 파라미터로 User를 받도록 수정

saint6839 committed:
refactor : 기존 primitive 에서 사용자 테이블과 연결되도록 변경

saint6839 committed:
refactor : 로그인 세션 정보 조건을 추가해서 기능 수정

-showQuestionForm(): 로그인 세션 정보가 있을 경우에만 질문 폼 띄워지도록 수정
-showEditForm(): 질문 수정폼 매핑
-create(): 사용자가 로그인했을 경우에만 질문 작성 가능하도록 숮어
-edit(): 작성자와 로그인한 사용자가 일치할 경우에만 질문내용 수정 가능하도록 수정
-remove(): 작성자와 로그인한 사용자가 일치할 경우에만 질문 삭제 가능하도록 수정

saint6839 committed:
feat: 예외 catch 할 수 있도록 추가

saint6839 committed:
chore: 불필요 import문 제거

Bellroute commented on src/main/java/codesquad/controller/QuestionController.java:
```diff
@@ -1,57 +1,112 @@
-        return nowDate + " " + formatedNowTime;
+    @GetMapping("/questions/form")
+    public String showQuestionForm(HttpSession session) {
+        Object value = session.getAttribute("sessionedUser");
```
타입과 변수명(심지어 타입없이도)만 봐도 의도를 알 수 있도록 코드를 변경해보면 어떨까요?

saint6839 replied:
지금보니 value라는게 너무 추상적으로 보이네요! 한번 고민해서 수정해보도록 하겠습니다 ㅎㅎ

Bellroute commented on src/main/java/codesquad/controller/QuestionController.java:
```diff
@@ -1,57 +1,112 @@
+
+    @GetMapping("/questions/{id}/editor")
+    public String showEditForm() {
+        return "qna/edit_form";
```
해당 question의 작성자가 아니어도 question 수정 화면에 접근 가능할 것 같은데 괜찮을까요?

saint6839 replied:
검증하는 로직이 빠져있네요..! 추가하도록 하겠습니다!

Bellroute commented on src/main/java/codesquad/controller/QuestionController.java:
```diff
@@ -1,57 +1,112 @@
+        if (value != null) {
+            Question question = new Question();
+            User sessionedUser = (User) value;
+            question.write(sessionedUser, writeQuestionDto, getWrittenTime());
```
write라는 메소드명이 의도가 모호하다고 생각이 들었습니다. 객체에 필요한 정보를 한번에 넘긴다는 점에서 생성자를 통해 정보를 전달하는 것과 유사해보이는데 어떠한 이점이 있을까요?

saint6839 replied:
기존에 생성자로 인자를 넘겨주는 방식에서 질문을 작성한다는 의미를 조금 더 보여주고 싶어서 메서드를 사용하게 되었던것 같습니다! 추가적으로 질문 수정 기능에서도 같은 구성의 폼을 사용하고 있기 떄문에, 수정 기능에서도 `write()` 메서드를 사용하고 있는데 개인적으로 이 부분은 같은 작성과 수정 모두 같은 기능을 가지고 있는데 수정 기능은 메서드를 다른 이름으로 만들어주어야 하는지 고민을 했던것 같습니다.

Bellroute commented on src/main/java/codesquad/controller/QuestionController.java:
```diff
@@ -1,57 +1,112 @@
+                Question question = questionRepository.findById(id)
+                                                      .filter(s -> s.getId()
+                                                                    .equals(sessionedUser.getId()))
+                                                      .orElseThrow(NoSuchUserException::new
```
학습한 내용을 적용해보려는 노력이 보이네요! 다만, 코드에서 의도한 것은 수정하려하는 question의 writer가 현재 세션 사용자와 다르기 때문에 Exception을 발생시키는 것인데, 사용자 없음을 나타내는 NoSuchUserException을 던지고 있는게 의도와는 조금 다른 결과를 리턴하는 것 같습니다.

또한, 함수형을 사용하는 이유 중 하나로 가독성을 꼽을 수 있을 것 같은데, 항상 함수형이 가독성이 좋은 것은 아니기 때문에 이 부분도 한번 고려해보시면 좋을 것 같습니다.

saint6839 replied:
앗.. 일단 있는 클래스 넣어놓자고 넣어두었던걸 수정을 안하고 그대로 PR에 올려버렸네요 ㅎㅎ.... 이전 혁진이형 리뷰에서 예외 클래스 부분 리뷰해주신 내용도 함께 고려해서 다음 리뷰때 반영해보도록 하겠습니다!

Bellroute commented on src/main/java/codesquad/controller/UserController.java:
```diff
@@ -1,60 +1,92 @@
-
-        editedUser.editInfo(user);
+    @PutMapping("/users/{id}")
+    public String editIndividualInfo(@PathVariable Long id, EditUserDto editUserDto) {
```
이 요청도 사용자 인증이 필요한 부분으로 보입니다.

Bellroute commented on src/main/java/codesquad/domain/Question.java:
```diff
@@ -1,34 +1,54 @@
+    private User writer;
+    private String title;
+    private String contents;
+    private LocalDateTime time;
```
java의 LocalDateTime은 데이터베이스에 들어갈 때 어떤 타입으로 들어갈까요?

saint6839 replied:
2024-09-09T10:00 와 같이 T가 붙은 형식으로 들어가게 됨을 확인하였습니다! 원하는 포맷으로 변경시켜주기 위해서 포매터의 `ofPattern()` 메서드를 이용해서 포매팅하여 사용하고 있습니다.

Bellroute commented on src/main/java/codesquad/domain/User.java:
```diff
@@ -1,37 +1,57 @@
+            this.name = editUserDto.getName();
+            this.email = editUserDto.getEmail();
+        } else {
+            throw new NoSuchPasswordException();
```
NoSuch\~보단 Wrong\~이나 Invalid\~ 등이 적절할 것 같은데 어떠신가요?

Bellroute commented on src/main/java/codesquad/domain/WriteQuestionDto.java:
```diff
@@ -0,0 +1,41 @@
+public class WriteQuestionDto {
+    private String title;
+    private String contents;
+    private LocalDateTime time;
```
작성하신 코드에서는 이 dto의 time 필드는 사용되지 않고 있는 것 같습니다. 추가하신 이유가 있을까요? + spring에서 요청 데이터를 
 LocalDateTime로 매핑하려면 프론트단에서 데이터를 어떠한 타입으로 넘겨야할까요?

Bellroute commented on src/main/java/codesquad/repository/QuestionRepository.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Optional;
+
+public interface QuestionRepository extends JpaRepository<Question, Long> {
+    Optional<Question> findByWriter(User writer);
```
해당 코드가 테스트 코드에서만 이용되었네요. 테스트 코드만을 위해서 작성되는 코드에 대해 한번 생각해보시면 좋을 것 같습니다.

saint6839 replied:
`findById()`로 테스트를 진행하려고 했었는데, `findById()`로 테스트를 진행하면 테스트 성공이 보장이 안되는 문제가 있어서 추가해주게 되었던 것 같습니다..! 아직 해결을 못하고 있는데 좀 더 찾아보고 수정해보도록 하겠습니다!

Bellroute commented on src/main/java/codesquad/controller/UserController.java:
```diff
@@ -1,60 +1,92 @@
+                                  .filter(s -> s.getPassword()
+                                                .equals(password))
+                                  .orElseThrow(NoSuchUserException::new);
+        session.setAttribute("sessionedUser", user);
```
현재는 "sessionedUser"를 key로 사용하고 있지만, 나중에 이 문자열을 변경할 일이 생긴다면 여러 군데에서 수정 작업이 필요해지고, 실수로 누락되는 부분이 생기면 문제가 생길 것 같습니다. 어떻게 리팩토링을 하면 좋을까요?

saint6839 replied:
공통 메서드로 분리해서 메서드를 통해 호출하도록 하면 좋을 것 같습니다!

Bellroute reviewed (commented):
수고 많으셨습니다!!
다른 분들께 말씀드린 것과 동일하게 Session에 대해서 더 학습해보시면 좋을 것 같습니다.

상엽님께서 테스트 코드도 작성하셔서 추가적으로 말씀드려봅니다! 
jpa에서 1차캐시, 더티체킹 등의 기능을 제공하고 있기 때문에, jpa를 포함한 테스트 코드를 작성하실 때는 쿼리 날라가는 것을 확인하셔서 의도하는대로의 테스트 환경에서 나온 결과인지 확인해볼 필요가 있습니다.
또한, 정상적인 요청에 정상적인 결과가 나오는지에 대한 테스트와 함께 비정상적인 요청에 의도한 결과가 나오는 지에 대한 테스트도 해보면 좋을 것 같습니다.
